test/aio5gc/context/ue.go

```diff
@@ -118,28 +111,43 @@ func (ue *UEContext) AddSmContext(newContext *SmContext) error {
 	defer ue.smContextMtx.Unlock()
 
 	sessionId := newContext.GetPduSessionId()
-	_, hasKey := ue.smContexts[sessionId]
+	oldContext, hasKey := ue.smContexts[sessionId]
 	if hasKey {
-		id := strconv.Itoa(int(sessionId))
-		return errors.New("[5GC] Could not create PDU Session " + id + " for UE " + ue.guti + ": already exist")
-	} else {
-		ue.smContexts[sessionId] = newContext
+		if !oldContext.state.Is(state.Inactive) {
+			id := strconv.Itoa(int(sessionId))
+			return errors.New("[5GC] Could not create PDU Session " + id + " for UE " + ue.guti + ": already in use")
+		}
 	}
+	ue.smContexts[sessionId] = newContext
 	return nil
 }
 
 func (ue *UEContext) DeleteSmContext(sessionId int32) (SmContext, error) {
+
+	var smContext SmContext
+	sc, err := ue.GetSmContext(sessionId)
+	if err != nil {
+		return SmContext{}, err
+	}
+	smContext = *sc
 	ue.smContextMtx.Lock()
 	defer ue.smContextMtx.Unlock()
+	delete(ue.smContexts, sessionId)
 
-	var smContext SmContext
+	return smContext, nil
+}
+
+func (ue *UEContext) GetSmContext(sessionId int32) (*SmContext, error) {
+	ue.smContextMtx.Lock()
+	defer ue.smContextMtx.Unlock()
+
+	var smContext *SmContext
 	_, hasKey := ue.smContexts[sessionId]
```

Comment:
smContext, hasKey := ue.smContexts[sessionId]

if !hasKey {
id := strconv.Itoa(int(sessionId))
return nil, errors.New("[5GC] Could not delete PDU Session " + id + " for UE " + ue.guti + ": not found")
}